src/states/node.rs

```diff
@@ -856,78 +851,72 @@ impl Node {
-                       peer_id);
+                       peer_id,
+                       error);
                 self.disconnect_peer(&peer_id);
             }
-            Some(AddedNodeDetails { must_notify, unneeded }) => {
-                info!("{:?} Added {:?} to routing table.", self, name);
+            Ok(split_prefix) => {
+                info!("{:?} Added {:?} to routing table.", self, public_id.name());
                 if self.peer_mgr.routing_table().len() == 1 {
                     self.send_event(Event::Connected);
                 }
-                for peer_id in self.peer_mgr.get_peer_ids(&must_notify) {
-                    let message = DirectMessage::NewNode(public_id);
-                    let _ = self.send_direct_message(&peer_id, message);
-                }
-                for peer_id in self.peer_mgr.get_peer_ids(&unneeded) {
-                    let message = DirectMessage::ConnectionUnneeded(*self.name());
-                    let _ = self.send_direct_message(&peer_id, message);
+
+                if let Some(prefix) = split_prefix {
+                    // None of these will have been in our group, so no need to notify Routing user.
+                    let peers_to_drop = self.peer_mgr.split_group(prefix);
+                    for peer_id in peers_to_drop {
+                        let _ = self.crust_service.disconnect(peer_id);
+                    }
                 }
 
-                self.reset_bucket_refresh_timer();
+                let all_rt_contacts = self.peer_mgr
+                    .routing_table()
+                    .iter()
+                    .filter(|name| *name != public_id.name())
+                    .cloned()
+                    .collect();
+                let peer_ids_to_notify = self.peer_mgr.get_peer_ids(&all_rt_contacts);
+                for to_notify in &peer_ids_to_notify {
+                    let message = DirectMessage::NewNode(public_id);
+                    let _ = self.send_direct_message(to_notify, message);
+                }
 
-                if common_groups {
-                    let event = Event::NodeAdded(name, self.peer_mgr.routing_table().clone());
+                if self.peer_mgr.routing_table().is_in_our_group(public_id.name()) {
+                    let message = DirectMessage::RoutingTable(self.peer_mgr
+                        .get_pub_ids(&all_rt_contacts));
+                    let _ = self.send_direct_message(&peer_id, message);
+                    let event = Event::NodeAdded(*public_id.name(),
+                                                 self.peer_mgr.routing_table().clone());
                     if let Err(err) = self.event_sender.send(event) {
                         error!("{:?} Error sending event to routing user - {:?}", self, err);
                     }
                 }
             }
         }
 
-        if self.peer_mgr.routing_table().len() == 1 {
-            self.request_bucket_close_groups();
-        }
```

Comment:
With the bucket requests gone, we currently rely exclusively on the peers handling the `NewNode` messages, and we don't ever fix the network if that is missed (e.g. because two nodes join at the same time), I think. That's fine for now, but we need to keep in mind that this is an issue: The consensus handling for the group member lists will need to make sure that all members of our own group _and_ all neighbouring groups eventually agree on the group constituency.